text/0000-overload-logic-operators.md

```diff
@@ -0,0 +1,211 @@
+
+This RFC also proposes to deprecate the `.or(...)`, `.or_with(|| ...)`, `.and(...)`, `.and_with(||
+...)`, `.unwrap_or(...)`, and `.unwrap_or_with(|| ...)` methods on `Option<T>` and `.or(...)`,
+`.or_with(|| ...)`, `.unwrap_or(...)`, and `.unwrap_or_with(|| ...)` methods on `Result<T, E>` since
+using this feature renders them unneeded.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The basis of this proposal are the two new traits. These should be implemented along the same lines
+as other operator traits. However, the desugaring should follow the `match` output in the previous
+section so as to obtain the desired short circuit operation.
+
+Once the traits have been implemented, several trait implementations should be added to the std library and the methods marked as deprecated.
+
+```rust
+impl LogicalOr<Option<T>> for Option<T> {
+    type Output = Self;
+    ...
+}
+
+impl trait LogicalOr<T> for Option<T> {
+    type Output = T;
+}
+
+impl trait LogicalAnd<Option<T>> for Option<T> {
+    type Output = Self;
+    ...
+}
+
+impl trait LogicalOr<Result<T, E>> for Result<T, E> {
+    type Output = Self;
+    ...
+}
+
+impl trait LogicalOr<T> for Result<T, E> {
+    type Output = T;
+}
+
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+1. Leaves the `||` and the `&&` as not strictly boolean operators, which might hurt readability
+2. Could lead to similarities to C++'s Operator bool() which => truthiness and is undesirable.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+- This design is the best because it does not rely on other traits, which are how the other operator traits work in Rust. It has the minimum overhead since it does not rely on closures.
+- Two alternatives were discussed and shown to be inferior in either performance, explicitness, or design.
+- The first being "truthiness" conversion trait and then automatically allowing `||` and `&&` if both that trait and the `BitOr` or `BitAnd` traits were also implemented. This was discarded because we did not want to go down the javascript route of auto converting things to boolean (Rust already does not allow non-bools in check expressions) and auto traits are not something that Rust has so that was another reason not to go down this route.
+- The second being a trait that accepts an `FnOnce` argument and then the second argument of the operator would be then implicitly hoisted into a closure. This was rejected both because hiding closures is not a zero-cost abstraction, it would break the similarity with the boolean operators because `None || return` would not return from the function unlike `false || return`. This also does not have any benifit over just using `or_with` directly except for a few characters.
+- If this is not done it then the usability of Rust without having to go to the docs would stay the same.
+
```

Comment:
This should mention the alternative suggested by @timvermeulen and @comex (adding an intermediate type); and if this RFC proposes not to do that, then it should list the reasons for not doing so.